### PR TITLE
Major rework of `optree_impl` and `dmtree_impl` for consistency.

### PR DESCRIPTION
**Pull request by hertschuh**
The `optree` implementation and the `dmtree` implementation of the `tree` API had a nummber of discreptancies. Running unit tests without `optree` installed would fail on a number of tests.

The documentation and behavior of the `tree` API was not internally consistent. There was contradicting documentation about the handling of `OrderedDict`s. The behavior of the `optree` implementation was to use the key sorted order for `pack_sequence_as` but use the \
sequence order for `flatten`, resulting in `flatten` + `pack_sequence_as` not being idempotent (as discovered in https://github.com/keras-team/keras/issues/20538 )

The exceptions used to report non-matching structures where different between the two implementation. When `optree` uses `ValueError` for all mimatches, `dmtree` would distinguish between `ValueError` and `TypeError` for some cases. This caused a number of bugs becau\
se `TypeError` was often not caught, only `ValueError`.

The `assert_same_structure` argument of `check_types` is deprecated and no longer does anything. The `optree` implementation would check the types of the *leaves*, whereas the `dmtree` would check the types of the *collections*. So `check_types=False` with `optree` w\
as fairly, although not completely similar, to `check_types=True` with `dmtree`. The rules is that no two collection types are considered the same, except for `dict`, `OrderedDict` and `defaultdict`.

Because `optree` is the default implementation used and `dmtree` is only a fallback, this PR changes the `tree` API behavior to conform to the `optree` approach everywhere. This makes the `optree` implementation a thin wrapper on top of `optree`, whereas large portio\
ns of the `dmtree` wrapper are now reimplemented in Python. Note that the `tree` API was initially modelled after `dmtree`, not `optree`.

There are a couple of fairly niche known discrepancies between the `optree` implementation and the `dmtree` implementation. They are documented in `dmtree_impl.py`.

- Fixed references to `unflatten_as` in documentation, which doesn't exist.
- Fixed contradicting documentation in `flatten` and `pack_sequence_as` related to the handling of `OrderedDict`s. The documentation now states that the sequence order is always used.
- Made handling of `OrderedDict`s follow the spec with both `optree` and `dmtree`.
- Made the exceptions raised consistent and documented them. `TypeError` is only for major programmer error (e.g. `func` is not callable), and `ValueError` is used for all structure mismatches.
- Removed all cases where `TypeError` was caught after a `assert_same_structure`.
- Fixed the discrepancy in the behavior for `namedtuple`s. The `optree` behavior is now followed, meaning that the path for fields are indices, not field names.
- Deprecated the `check_types` argument in `assert_same_structure` and implemented the `optree` behavior in `dmtree`.
- Remove `sequence_fn` argument	of `pack_sequence_as`, which was not used and force the	`optree` implementation	to be fully rewritten in Python.
- Added `MAP_TO_NONE` to the API, added support for it in both implementations of `traverse`. This feature was documented, but not accessible and not actually implemented.
- Added support for registered classes with `dmtree`, both with `flatten` and `unflatten` passed at registration time and methods on the class.
- Tracked collections are now supported with `dmtree` (`TrackedList`, `TrackedSet` and `TrackedDict`). In particular, `TrackedSet` would be handled as a leaf and never traversed.
- Removed dependency of tracked collections on `optree` in `tree_flatten` and `tree_unflatten`.
- Tensorflow's `ListWrapper` and `_DictWrapper` are now supported with `dmtree`.
- Implemented a more efficient way for `optree` to verify structures are the same while traversing them with `map_structure` and `map_structure_up_to`. This avoids multiple traversals.
- Added documentation for `list_to_tuples` and `map_shape_structure`.
- Completely rewrote all tree API unit tests, which are now painfully thorough.
- `map_shape_structure` is now unit tested.
- Fixed unintented use of `tree` instead of `keras.tree` in unit test.
- Ran unit tests for all backends with `optree` uninstalled.

Fixes https://github.com/keras-team/keras/issues/20538

### CODE REVIEW

File: keras/src/trainers/compile_utils_test.py [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
 import numpy as np
-import tree
 from absl.testing import parameterized
 
 from keras.src import backend
 from keras.src import metrics as losses_module
 from keras.src import metrics as metrics_module
 from keras.src import ops
 from keras.src import testing
+from keras.src import tree
```
Comment (fchollet) [inline keras/src/trainers/compile_utils_test.py:11]: This seems to be triggering a failure on the torch side https://github.com/keras-team/keras/actions/runs/11785192949/job/32825931002?pr=20481

(inability to recognize a NamedTuple as a loss tuple)

**Review by fchollet — COMMENTED** [review]
Thanks for the fix!

File: keras/src/trainers/compile_utils_test.py (hunk 1 above)
Comment (hertschuh) [reply keras/src/trainers/compile_utils_test.py:11]: Looking into it. For some reason, this test required `dm-tree`.

Testing further, I tested with `optree` (the default) uninstalled and before this change, and a number of tests fail.

So some tests (and maybe some code) are still tree implementation dependent.

**Comment by nicolaspi** [comment]
Wouldn't it be better to wrap `OrderedDict` rather than re-implement `flatten`, which is written in C++ in optree?

e.g.:

```python
from collections import OrderedDict
import optree
from keras import tree

class WrappedOrderedDict(OrderedDict):
  pass

def flatten(d):
  values = []
  keys = []
  for key in sorted(d.keys()):
    values.append(d[key])
    keys.append(key)
  return values, list(d.keys()), keys

def unflatten(metadata, children):
  index = {key: i for i, key in enumerate(sorted(metadata))}
  return OrderedDict({key: children[index[key]] for key in metadata})


optree.register_pytree_node(
    WrappedOrderedDict,
    flatten,
    unflatten,
    namespace="keras",
)


def ordereddict_pytree_test():
  # Create an OrderedDict with deliberately unsorted keys
  ordered_d = OrderedDict([('c', 3), ('a', 1), ('b', 2)])
  
  def wrap(s):
    if isinstance(s, OrderedDict):
      return WrappedOrderedDict(s)
    return None

  def unwrap(s):
    if isinstance(s, WrappedOrderedDict):
      return OrderedDict(s)
    return None

  d = tree.traverse(wrap, ordered_d, top_down=False)
  flat_d = tree.flatten(d)
  flat_d_paths = tree.flatten_with_path(d)
  assert flat_d == [1, 2, 3]
  assert [p[0] for p, v in flat_d_paths] == ["a", "b", "c"]
  
  tree_struct = tree.traverse(wrap, ordered_d, top_down=False)
  wrapped_d = tree.pack_sequence_as(tree_struct, flat_d)
  orig_d = tree.traverse(unwrap, wrapped_d, top_down=False)
  assert isinstance(orig_d, OrderedDict)
  assert list(orig_d.keys()) == list(ordered_d.keys())
  assert list(orig_d.values()) == list(ordered_d.values())


ordereddict_pytree_test()
```

**Comment by hertschuh** [comment]
> Wouldn't it be better to wrap `OrderedDict` rather than re-implement `flatten`, which is written in C++ in optree?

Hi Nicolas,

Thank you for the suggestion. I actually completely scratched this PR and decided to use a different approach. The `optree` behavior will be the reference behavior. The goal is indeed to maximize the use of the C++ implementation of `optree` since it is the default and `dm-tree` is only a fallback.

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the contribution! 👍
